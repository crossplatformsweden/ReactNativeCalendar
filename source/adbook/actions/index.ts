import * as Redux from 'redux';
import * as types from '../types';

export const Book = (book: types.IBooking, bookingaction: types.ParkingConstants) => async (
    dispatch: Redux.Dispatch<types.IParkingAction>
) => {
        const booking = new types.Booking(
            book.timeperiod,
            book.location,
            book.vehicle,
            book.adress,
            book.lessee
        );

        const emptybooking = new types.Booking(
            null, null, null, null, null
        );

        const changedbooking = new types.Booking(
            booking.timeperiod,
            booking.location,
            booking.vehicle,
            booking.adress,
            booking.lessee
        );

        if (bookingaction === types.ParkingConstants.BOOKING_ADDED) {

            const theBooking: types.IParkingAction = {
                type: types.ParkingConstants.BOOKING_ADDED,
                booking: booking,
              };

            dispatch(theBooking);
            return;
        }

        if (bookingaction === types.ParkingConstants.BOOKING_CANCELLED) {

            const theBooking: types.IParkingAction = {
                type: types.ParkingConstants.BOOKING_CANCELLED,
                booking: emptybooking,
              };

            dispatch(theBooking);
            return;
        }

        if (bookingaction === types.ParkingConstants.BOOKING_CHANGED) {
            const theBooking: types.IParkingAction = {
                type: types.ParkingConstants.BOOKING_CHANGED,
                booking: changedbooking,
              };
            dispatch(theBooking);
        }

};