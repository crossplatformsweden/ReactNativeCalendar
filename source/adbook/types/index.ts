import { Book } from '../actions';
import Moment from 'moment';

export enum ParkingConstants {
    BOOKING_ADDED = 'BOOKING_ADDED',
    BOOKING_CHANGED = 'BOOKING_CHANGED',
    BOOKING_CANCELLED = 'BOOKING_CANCELLED',
}

export type Book = typeof Book;

export enum IVehicleType {
    Van= 'Van',
    Truck= 'Truck',
    Sedan = 'Sedan',
    SUV= 'SUV',
    /**
     * AKA Family Car, SV: Kombi
     *
     * @type {'StationWagon'}
     * @memberof IVehicleType
     */
    StationWagon= 'StationWagon',
}

interface IVehicle {
    vehicleType: IVehicleType;
    vehicleModel?: string;
    regNumber: string;
}

interface IPerson {
    firstName: string;
    lastName: string;
}

interface ITimePeriod {
    timeStart: Moment.Moment;
    timeEnd: Moment.Moment;
}
interface ILocation {
    lat: number;
    lon: number;
}

export interface IBooking {
    timeperiod: ITimePeriod;
    location: ILocation;
    vehicle: IVehicle;
    adress: string;
    lessee: IPerson;
}

export interface IBookingState {
    booking: IBooking;
}

export interface IParkingAction extends IBookingState {
    type: ParkingConstants;
}

export class Booking implements IBooking {
    constructor (readonly timeperiod: ITimePeriod, readonly location: ILocation,
        readonly vehicle: IVehicle, readonly adress: string, readonly lessee: IPerson ) {

        }
}
