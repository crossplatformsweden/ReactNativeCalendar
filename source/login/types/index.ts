// login/types/index
import { FacebookLogin, GoogleLogin, BankIdLogin } from '../actions';

export type FacebookLogin = typeof FacebookLogin;

export type GoogleLogin = typeof GoogleLogin;

export type BankIdLogin = typeof BankIdLogin;

/**
 * A user object created when logging in
 * @typedef IUser
 * @type {Object}
 * @public
 * @export
 * @property {string} accessToken - social provider access token
 * @property {string} name - full name of the user
 * @property {string} picture - profile picture url if any
 * @property {string} type - Facebook, Google etc
 * @public
 */
export interface IUser {
  accessToken: string;
  name: string;
  picture?: string;
  type: string;
  username?: string;
  password?: string;
  email?: string;
  autologin?: boolean;
}

export class User implements IUser {
  constructor(
    readonly accessToken: string,
    readonly name: string,
    readonly picture: string = '',
    readonly type: string,
    readonly username: string = null,
    readonly password: string = null,
    readonly email: string = null,
    readonly autologin: boolean = false
  ) {}
}

/**
 * LoginReducer constant string keys
 *
 * @namespace
 * @public
 * @typedef LoginConstants
 * @type {Object}
 */
export enum LoginConstants {
  /**
   * Login failed or was cancelled
   */
  LOGIN_FAILED = 'LOGIN_FAILED',
  /**
   * Login state: authentication succeeded
   */
  LOGIN_SUCCESS = 'LOGIN_SUCCESS',

  /**
   * Login state: authentication in progress
   */
  LOGIN_BUSY = 'Kollar vem du Ã¤r',

  /**
   * Login state: logged out of app
   */
  LOGGED_OUT = 'LOGGED_OUT',
}

export interface ILoginState {
  /**
   * The user payload recived from login if any
   *
   * @type {IUser}
   * @memberof ILoginReducer
   */
  user?: IUser;
  /**
   * Determines if the user is successfully logged in
   *
   * @type {boolean}
   * @memberof ILoginReducer
   */
  isLoggedIn: boolean;
}

export interface ILoginAction extends ILoginState {
  type: LoginConstants;
}
