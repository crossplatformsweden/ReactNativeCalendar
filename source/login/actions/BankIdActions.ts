// login/actions/GoogleActions
import * as Redux from 'redux';
import * as login from '../types/';

import { SaveByKey } from '../../storage';
import { AppLoadingChanged, AppErrorChanged } from '../../utility';
import { StorageConstants } from '../../storage/types';
import { WebBrowser } from 'expo';

/**
 * The resulting user
 * @type {User}
 * @memberof GoogleActions
 */
let user: login.IUser = null;

const bankIdLogin = 'BankIdLogin';

/**
 * Login with European BankId
 * @function BankIdLogin
 * @returns {UserDispatch}
 * @public
 * @global
 */
// @ts-ignore
export const BankIdLogin = (personalNr: string) => async (
  dispatch: Redux.Dispatch<login.ILoginAction>
) => {
  dispatch(
    AppLoadingChanged({
      isBusy: true,
      reason: login.LoginConstants.LOGIN_BUSY,
      sender: bankIdLogin,
    })
  );

  try {
    // Call bankid
    const result = await WebBrowser.openBrowserAsync(
      'bankid:///?autostarttoken=a4904c4c-3bb4-4e3f-8ac3-0e950e529e5f&' +
      'redirect=https%3a%2f%2fdemo.bankid.com%2fnyademobanken%2fCavaClientRedirReceiver.aspx%3forder' +
      'Ref%3dbedea56d-7b46-47b1-890bf787c650bc93%26returnUrl%3d.%2fCavaClientAuth.aspx%26Environment%3dKundtest'
    );

    // @ts-ignore
    if (result.type === 'success') {
      user = new login.User(
        'result.idToken',
        'result.user.name',
        'result.user.photoUrl',
        'BankId'
      );

      dispatch({
        type: login.LoginConstants.LOGIN_SUCCESS,
        payload: user,
      });

      // Trigger Save to storage action
      await dispatch(SaveByKey(StorageConstants.STORAGE_USER_KEY, user));

      // Reset errors
      dispatch(AppErrorChanged({ hasError: false, sender: bankIdLogin }));
    } else {
      // Cancelled
      dispatch(
        AppErrorChanged({
          hasError: true,
          reason: 'Login was cancelled',
          exception: new Error(login.LoginConstants.LOGIN_FAILED),
          sender: bankIdLogin,
        })
      );
    }
  } catch (error) {
    dispatch(
      AppErrorChanged({
        hasError: true,
        reason: login.LoginConstants.LOGIN_FAILED,
        exception: error,
        sender: bankIdLogin,
      })
    );
  } finally {
    // Reset busy
    dispatch(AppLoadingChanged({ isBusy: false, sender: bankIdLogin }));
  }
};
