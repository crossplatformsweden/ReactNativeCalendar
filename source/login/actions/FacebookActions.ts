// login/actions/index
import * as Redux from 'redux';
import axios from 'axios';
import { Constants, Facebook } from 'expo';

import * as loginTypes from '../types/';
import { AppErrorChanged, AppLoadingChanged } from '../../utility/';
import { StorageTypes, SaveByKey } from '../../storage';

/**
 * The resulting user
 * @type {User}
 * @memberof FacebookActions
 */
let user: loginTypes.IUser = null;

const facebookLogin = 'FacebookLogin';

/**
 * Login to Facebook and store user to AsyncStorage
 * @function FacebookLogin
 * @returns {UserDispatch}
 * @memberof FacebookActions
 * @public
 * @global
 */
export const FacebookLogin = () => async (
  dispatch: Redux.Dispatch<loginTypes.ILoginAction>
) => {
  dispatch(
    AppLoadingChanged({
      isBusy: true,
      reason: loginTypes.LoginConstants.LOGIN_BUSY,
      sender: facebookLogin,
    })
  );

  try {
    const { type, token } = await Facebook.logInWithReadPermissionsAsync(
      Constants.manifest.facebookAppId,
      {
        behavior: Constants.appOwnership === 'standalone' ? 'native' : 'web',
      }
    );

    if (type === 'success' && token) {
      // Get the user's name using Facebook's Graph API
      const graphRequest = `https://graph.facebook.com/me?fields=id,name,picture&access_token=${token}`;
      console.log('FACEBOOK GRAPH API REQUEST');
      console.log(graphRequest);
      const response = await axios.get(
        graphRequest
      );

      /** @type {FacebookActions.User} */
      user = new loginTypes.User(
        token,
        response.data.name,
        response.data.picture.data.url,
        'Facebook'
      );

      const resultAction: loginTypes.ILoginAction = {
        type: loginTypes.LoginConstants.LOGIN_SUCCESS,
        user,
        isLoggedIn: true,
      };

      dispatch(resultAction);

      // Trigger Save to storage action
      await dispatch(
        SaveByKey(StorageTypes.StorageConstants.STORAGE_USER_KEY, user)
      );

      // Reset errors
      dispatch(AppErrorChanged({ hasError: false, sender: facebookLogin }));
    } else if (type === 'cancel') {
      dispatch(
        AppErrorChanged({
          hasError: true,
          reason: 'Login was cancelled',
          exception: new Error(loginTypes.LoginConstants.LOGIN_FAILED),
          sender: facebookLogin,
        })
      );
    }
  } catch (error) {
    dispatch(
      AppErrorChanged({
        hasError: true,
        reason: loginTypes.LoginConstants.LOGIN_FAILED,
        exception: error,
        sender: facebookLogin,
      })
    );
  } finally {
    dispatch(AppLoadingChanged({ isBusy: false, sender: facebookLogin }));
  }
};

export { axios };
