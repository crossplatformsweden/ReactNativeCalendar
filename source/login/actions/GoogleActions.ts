// login/actions/GoogleActions
import { Google, Constants } from 'expo';
import * as Redux from 'redux';
import * as loginTypes from '../types/';

import { AppLoadingChanged, AppErrorChanged } from '../../utility';
import { StorageTypes, SaveByKey } from '../../storage';

// Remarks: native will only work on standalone deployment
const config =
  Constants.appOwnership === 'standalone'
    ? Constants.platform.ios
      ? {
          iosStandaloneAppClientId:
            Constants.manifest.extra.google.iosStandaloneAppClientId,
        }
      : {
          androidStandaloneAppClientId:
            Constants.manifest.extra.google.androidStandaloneAppClientId,
        }
    : {
        androidClientId: Constants.manifest.extra.google.androidClientId,
        iosClientId: Constants.manifest.extra.google.iosClientId,
        webClientId: Constants.manifest.extra.google.webClientId,
      };

/**
 * The resulting user
 * @type {User}
 * @memberof GoogleActions
 */
let user: loginTypes.IUser = null;

const methodName = 'GoogleLogin';

/**
 * Login to Google and store user to AsyncStorage
 * @function GoogleLogin
 * @returns {UserDispatch}
 * @memberof GoogleActions
 * @public
 * @global
 */
export const GoogleLogin = () => async (
  dispatch: Redux.Dispatch<loginTypes.ILoginState>
) => {
  dispatch(
    AppLoadingChanged({
      isBusy: true,
      reason: loginTypes.LoginConstants.LOGIN_BUSY,
      sender: methodName,
    })
  );

  try {
    const result = await Google.logInAsync(config);
    if (result.type === 'success') {
      /** @type {GoogleActions.User} */
      user = new loginTypes.User(
        result.idToken,
        result.user.name,
        result.user.photoUrl,
        'Google'
      );

      dispatch(
        loginTypes.LoginState(
          loginTypes.LoginConstants.LOGIN_SUCCESS,
          user,
          true
        )
      );

      // Trigger Save to storage action
      await dispatch(
        SaveByKey(StorageTypes.StorageConstants.STORAGE_USER_KEY, user)
      );

      // Reset errors
      dispatch(AppErrorChanged({ hasError: false, sender: methodName }));
    } else {
      dispatch(
        loginTypes.LoginState(
          loginTypes.LoginConstants.LOGIN_CANCELLED,
          null,
          false
        )
      );
    }
  } catch (error) {
    dispatch(
      AppErrorChanged({
        hasError: true,
        reason: loginTypes.LoginConstants.LOGIN_FAILED,
        exception: error,
        sender: methodName,
      })
    );
  } finally {
    dispatch(AppLoadingChanged({ isBusy: false, sender: methodName }));
  }
};
