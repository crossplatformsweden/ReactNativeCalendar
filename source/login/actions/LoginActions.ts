// login/actions/LoginActions
import { Actions } from 'react-native-router-flux';
import * as Redux from 'redux';

import Store from '../../Store';
import * as login from '../types/';
import * as loginTypes from '../../login/types';
import { GetByKey, RemoveKey, StorageTypes } from '../../storage';
import { AppLoadingChanged, AppErrorChanged } from '../../utility';

const className = 'LoginActions';

/**
 * Try to automatically login the user by getting the user from storage
 *
 * Will dispatch login state as "logged in" if user exist
 */
export const AutoLogin = () => async (
  dispatch: Redux.Dispatch<login.ILoginState>
) => {
  const userState = Store.getState().login;
  if (userState.user && userState.isLoggedIn) {
    // Already logged in
    return;
  }

  try {
    // Get user from storage
    dispatch(
      AppLoadingChanged({
        isBusy: true,
        reason: login.LoginConstants.LOGIN_BUSY,
        sender: className,
      })
    );

    // Get user from storage
    await dispatch(GetByKey(StorageTypes.StorageConstants.STORAGE_USER_KEY));

    const newUser = Store.getState().storage.value as loginTypes.IUser;

    if (newUser) {
      // Report user as logged in
      const newLoginState = loginTypes.LoginState(
        login.LoginConstants.LOGIN_SUCCESS,
        newUser,
        true
      );

      dispatch(newLoginState);

      // Reset errors
      dispatch(AppErrorChanged({ hasError: false, sender: className }));
    }
  } catch (error) {
    dispatch(
      AppErrorChanged({
        hasError: true,
        reason: login.LoginConstants.LOGIN_FAILED,
        exception: error,
        sender: className,
      })
    );
  } finally {
    dispatch(AppLoadingChanged({ isBusy: false, sender: className }));
  }
};

export type AutoLogin = typeof AutoLogin;

/**
 * Logout the user by removing from state and navigating to login screen
 */
export const Logout = () => async (
  dispatch: Redux.Dispatch<login.ILoginState>
) => {
  // Remove user from storage
  await dispatch(RemoveKey(StorageTypes.StorageConstants.STORAGE_USER_KEY));

  // Set state to logged out
  dispatch(
    loginTypes.LoginState(loginTypes.LoginConstants.LOGGED_OUT, null, false)
  );

  // Display login screen
  Actions.LoginScreen();
};

export type Logout = typeof Logout;
