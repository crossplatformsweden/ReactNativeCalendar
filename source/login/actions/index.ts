import { FacebookLogin } from './FacebookActions';
import { GoogleLogin } from './GoogleActions';
import { BankIdLogin } from './BankIdActions';
import { AutoLogin } from './LoginActions';

export { FacebookLogin, GoogleLogin, BankIdLogin, AutoLogin };
