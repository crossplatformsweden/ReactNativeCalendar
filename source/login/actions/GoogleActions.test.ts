// login/actions/index.test
// @ts-ignore
/// <reference types="jest"/>
import configureMockStore from 'redux-mock-store';
import AxiosMockFactory from 'axios-mock-adapter';

import { Store } from '../../Store';
import * as loginTypes from '../types';
import * as types from '../../Types';
import { GoogleLogin } from '../actions/GoogleActions';
import { UtilityTypes } from '../../utility/';
import * as StorageTypes from '../../storage/types';

/**
 * Mock Redux store
 */
const store = configureMockStore<Store>();

// Navigator will call connected components
jest.mock('../../navigator', () => ({
  Routes: 'View',
  AppNavigator: 'Button',
}));

describe('Google actions', () => {
  beforeEach(() => {
    // @ts-ignore
    store.clearActions();
  });

  test('GoogleLogin returns SUCCESS when done', async () => {
    const expectedPayload = new loginTypes.User(
      'id',
      'test',
      'http://www.crossplatform,se/',
      'Google'
    );

    const methodName = 'GoogleLogin';
    const methodNameStorageSave = 'SaveByKey';

    const expectedActions = [
      UtilityTypes.AppLoadingChangedAction(
        UtilityTypes.UtilityConstants.APP_LOAD_BUSY,
        loginTypes.LoginConstants.LOGIN_BUSY,
        methodName
      ),
      loginTypes.LoginState(
        loginTypes.LoginConstants.LOGIN_SUCCESS,
        expectedPayload,
        true
      ),
      UtilityTypes.AppLoadingChangedAction(
        UtilityTypes.UtilityConstants.APP_LOAD_BUSY,
        StorageTypes.StorageConstants.STORAGE_BUSY,
        methodNameStorageSave
      ),
      StorageTypes.StorageState(
        StorageTypes.StorageConstants.STORAGE_SAVED,
        StorageTypes.StorageConstants.STORAGE_USER_KEY,
        expectedPayload
      ),
      UtilityTypes.AppErrorChangedAction(
        UtilityTypes.UtilityConstants.APP_NO_ERROR,
        null,
        null,
        methodNameStorageSave
      ),
      UtilityTypes.AppLoadingChangedAction(
        UtilityTypes.UtilityConstants.APP_LOAD_DONE,
        null,
        methodNameStorageSave
      ),
      UtilityTypes.AppErrorChangedAction(
        UtilityTypes.UtilityConstants.APP_NO_ERROR,
        null,
        null,
        methodName
      ),
      UtilityTypes.AppLoadingChangedAction(
        UtilityTypes.UtilityConstants.APP_LOAD_DONE,
        null,
        methodName
      ),
    ];

    // Dispatch action
    // @ts-ignore
    await store.dispatch(GoogleLogin());

    // @ts-ignore
    expect(store.getActions()).toMatchSnapshot();
    // @ts-ignore
    expect(store.getActions()).toEqual(expectedActions);
  });
});
