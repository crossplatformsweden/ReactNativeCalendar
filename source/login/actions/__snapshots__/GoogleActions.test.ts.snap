// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Google actions GoogleLogin returns SUCCESS when done 1`] = `
Array [
  Object {
    "reason": "Kollar vem du Ã¤r",
    "sender": "GoogleLogin",
    "type": "APP = BUSY LOADING",
  },
  Object {
    "isLoggedIn": true,
    "type": "LOGIN_SUCCESS",
    "user": User {
      "accessToken": "id",
      "autologin": false,
      "email": null,
      "name": "test",
      "password": null,
      "picture": "http://www.crossplatform,se/",
      "type": "Google",
      "username": null,
    },
  },
  Object {
    "reason": "Friskar upp minnet",
    "sender": "SaveByKey",
    "type": "APP = BUSY LOADING",
  },
  Object {
    "key": "STORAGE KEY = USER",
    "type": "STORAGE = SAVED KEY",
    "value": User {
      "accessToken": "id",
      "autologin": false,
      "email": null,
      "name": "test",
      "password": null,
      "picture": "http://www.crossplatform,se/",
      "type": "Google",
      "username": null,
    },
  },
  Object {
    "exception": null,
    "reason": null,
    "sender": "SaveByKey",
    "type": "APP NO ERROR",
  },
  Object {
    "reason": null,
    "sender": "SaveByKey",
    "type": "APP = DONE LOADING",
  },
  Object {
    "exception": null,
    "reason": null,
    "sender": "GoogleLogin",
    "type": "APP NO ERROR",
  },
  Object {
    "reason": null,
    "sender": "GoogleLogin",
    "type": "APP = DONE LOADING",
  },
]
`;
